pkg/runner/runner.go

```diff
@@ -169,13 +169,13 @@ func (runner *runnerImpl) NewPlanExecutor(plan *model.Plan) common.Executor {
 					})
 					b++
 					if b == maxParallel {
-						pipeline = append(pipeline, common.NewParallelExecutor(stageExecutor...))
+						pipeline = append(pipeline, common.NewParallelExecutor(maxParallel, stageExecutor...))
 						stageExecutor = make([]common.Executor, 0)
 						b = 0
 					}
 				}
 			}
-			return common.NewPipelineExecutor(pipeline...)(ctx)
+			return common.NewParallelExecutor(runtime.NumCPU(), pipeline...)(ctx)
```

Comment:
Should probably be configurable via CLI switch (a la `make -j` and friends), and default to `runtime.NumCPU() - 1` to prevent complete system crashes. (I've had a few such crashes, myself, over the last week, so I'm particularly sensitive to them at the moment.)

Reply:
Yes, thought about making this configurable as well.
We are already overcomming here, since the limit to num-cpu applies only to parallel jobs.
We would spawn 4x that number if all jobs have matrices.

Reply:
I don't think this could, by default, lead to a system crash.
If you imagine a computer with 4 CPUs, we will at most run 16 goroutines. Unless each of them are burning a CPU completely, this should be okay.
Even saving a CPU by reducing parallel tasks does not help then, since that would still be 12 goroutines.
But with 'max-parallel' on the task level it could be reduced to 1 per talk which would be totally fine.

So I think the configurable limit could be done in another PR.

Reply:
There is no point to introduce additional complexity of limiting CPU used by act and making new flags. We already provide options to choose specific jobs/workflows.

Another thing is that this CPU limit is not that important at all as CPU scheduler will handle those containers just fine.
What you should be afraid of is what is running in the workflow (e.g. build Rust crate and you will see)

I've never seen a system OS "crash" due to CPU overcommitting.

Reply:
Choosing specific workflows and jobs doesn't choose specific matrices, and anyway only running specific matrices wouldn't solve the larger issue of trying to run an entire workflow locally without overloading the entire host system in the process. As you pointed out, it's the jobs inside the containers that cause problems, and act doesn't have any reasonable way to control that. But it does have a reasonable way to control itself, and letting the user control how many jobs to run in parallel would let full workflow runs with multiple parallel builds/tests be feasible on hardware that isn't a data center like GHA itself.

Reply:
It is perfectly possible to set `max-parallel` in your matrices jobs to `1` and act will run at most one job per CPU in your system. That should be fine for 99.99% of all cases.

Reply:
You can chip in here: https://github.com/nektos/act/issues/114
Specifying CPU limit is waste of time and resources

Reply:
For that usecase you could try my act clone https://github.com/ChristopherHX/runner.server. In my opinion it is more feature complete than nektos/act.
The `Runner.Client` or `gharun` executable is an act replacement
It has similar cli switches than `nektos/act`, but is based on actions/runner c#.
It has a `--parallel` switch for selecting how many actions/runner are spawned on your pc. ( defaults to 1 )
It has a `--matrix` switch for selecting a specfix matrix entry of a job, but then you won't be able to run all jobs of your workflow.
However it has a performance problem if you aren't on linux and use linux containers ( on windows you should call it from a windows subsystem for linux shell )

Reply:
> It is perfectly possible to set `max-parallel` in your matrices jobs to `1`

The idea is to let GHA run things in parallel while my local system can't. Setting `max-parallel` to anything defeats that purpose.

> You can chip in here: #114

Saw that, interacted with existing comments, will be useful to have when fixing specific issues. Doesn't really address the opposite issue with running the entire workflow, though.

> Adding option to specify CPU limit is waste of time and resources

Not trying to limit CPU, but rather parallel execution. This PR already does exactly that, as does its predecessor for matrices. The switch i was requesting was to manually override the default number used as the limiter, whatever that might be. Some folks will have workflows that don't need limiting at all, and others will have systems that weren't made in the last two years and need to limit things further than the default. I recognize that i mentioned a minus-one approach in my original review, but that's entirely separate from the switch I'd like to see, and honestly irrelevant to anything, as pointed out above in responses to that review.

> For that usecase you could try my act clone 

I'll have a look. I don't need the `--matrix` switch as often as the `--parallel` one, which is why I commented here and not over on 114, but having both is nice. I imagine I'll have to alter my workflows for the implementation differences, though, which is why I was trying to stick with act if i can, since it already powers the majority of my workflows in other repos. It's just when i have to build large numbers of packages in parallel that things start to go sideways, and the ability to have more control over how act spins those tasks up becomes important rather than overkill. I can't exactly change my tech stack between local and GHA without defeating the entire purpose of a local run.

Reply:
> The idea is to let GHA run things in parallel while my local system can't. Setting max-parallel to anything defeats that purpose.

You can use expressions and the `env` context with the value of `ACT` to decide if you are running locally (where you have more limited concurrency) compared to GHA where you have a farm of computers.

Others that that, a more complex implementation with a global limit which could be obeyed is probably welcome.
This PR allows to have concurrency at all, which is not defined by matrices.

Reply:
Right. The matrices bit was a side note.